File: src/main/java/me/creepinson/creepinoutils/util/upgrade/IUpgradeable.java
package me.creepinson.creepinoutils.util.upgrade;

import net.minecraft.item.ItemStack;

import java.util.List;

/**
 * @author Theo Paris https://theoparis.com
 * Project creepinoutils
 **/
public interface IUpgradeable {
    boolean upgrade(UpgradeInfo info);

    ItemStack removeUpgrade(Upgrade upgrade);

    boolean canUpgrade();

    List<UpgradeInfo> getStoredUpgrades();

}

File: src/main/kotlin/com/theoparis/creepinoutils/util/upgrade/IUpgradeable.kt
package com.theoparis.creepinoutils.util.upgrade

import net.minecraft.item.ItemStack

interface IUpgradeable {
    fun upgrade(info: UpgradeInfo): Boolean
    fun removeUpgrade(upgrade: Upgrade): ItemStack
    fun canUpgrade(): Boolean
    val storedUpgrades: List<Any?>
}